### DISCUSSION TITLE
Uso correto do `if`

### DISCUSSION THREAD

**Discussion by rubenmarcus** [frontendbr/forum#296] [Discussão]
Oi, li um artigo esses dias que um dev publicou no face falando sobre ifs hadouken.
https://diogommartins.wordpress.com/2016/08/22/como-se-defender-de-ifs-hadouken/

Esses dias um dev que trabalha comigo disse que se você usa else, é porque seu if está mal estruturado.
Tambem me disse que else if é uma coisa abominavel.

Qual o correto? Digo em questão de lógica de programação e melhora de código?

Pra mim ( minha opinião ) é que se a linguaguem que você está utilizando tem um recurso, e ele atende sua necessidade no momento, você deve usar. 
E por necessidade eu não digo o padrão ideal de deixar um código enxuto e performatico, as vezes fazer um código ruim e redundante, supre uma necessidade imediata, que você ia gastar um pouco mais de tempo com lógica ou claro se você fosse um programador melhor. (deixando bem claro que eu nunca fiz um if haduken, mas sendo sincero ja fiz if dentro de if - 2 camadas)

**Comment by gabrieldarezzo** [comment]
Sou totalmente a favor da descomplicar a vida....
**Return early** é algo que sigo no dia a dia e acredito que simplifica o código de uma forma muito eficaz.

Sobre estar utilizando o if e ser incorreto  seria algo relacional ao paradigma de  O.O. ou da programação funcional será?

É gratificante o código 'rodar pelo computador.'
Mas alguém (outro humano) entender o que você pretendia com aquele código é melhor ainda xD

**Comment by mtmr0x** [comment]
Quem define uso correto de `if` é equipe e contexto, assim como "callback hell", a estrutura visual de um `if hadouken` é a mesma.

Eu diria que o maior erro é criar uma estrutura grande de `if`, onde a condicional verdadeira ou falsa executaria uma série de coisas. Os paradigmas O.O. e funcional em relação a isso, que o @gabrieldarezzo comentou, não tem ninguém falando para não usar ou usar, pois o uso do `if` é tipo viver em sociedade, tem umas regras de respeito geral e ética e assim que deve ser, em todas linguagens.

As "vantagens", a super grosso modo falando, de O.O. com estruturas de `if` e `else` é como tu és permitido a levantar os níveis de lógica fazendo condicionais simples e trazendo ações para o baixo nível apenas executar e o alto nível ir trabalhando em tornar o objeto mais simples. 

Não sei se sabe bem sobre esse negocio de high level pra low level, veja isto (caso conheça, ignore e desculpe-me): https://en.wikipedia.org/wiki/High-_and_low-level
(tem exemplos de uso para a terminologia, que tu pode abstrair e entender, pois muita coisa em programação não é exclusiva desse mundo 🙈 ).

Em programação funcional vem as abstrações em cima de abstrações, que se tu passar para uma equação matemática, seria uma forma um tanto diferente de escrever do paradigma de orientação a objeto, com o mesmo resultado e a grande intenção é abstrair a lógica nos lugares certos para, por mais que mais extenso seja as viagens entre arquivos, seja mais simples de compreender o que está acontecendo.

RESUMO DO TEXTÃO: `if` é uma coisa que tu vai usar muito na vida e quanto melhor programador tu se tornar, entendendo linguagens e conceitos, principalmente pensando em mantenabilidade, melhor tu saberás o uso.

Desculpa o textão, espero ter ajudado.

**Comment by fdaciuk** [comment]
Não diria que existe um "uso correto", mas sim o uso que melhor se encaixa naquilo que você quer dizer com seu código.

Usar if / else / elseif não é um problema. É melhor evitar else / elseif? Sim, pois deixa o seu código menos complexo. Mas tem momentos que usar else pode se tornar mais legível do que evitá-lo.

Minha recomendação seria: Use. Se ficou difícil de entender, e for possível, use ternário. Se ainda ficou ruim, tente ser mais específico: quebre a lógica que vai dentro do if em uma função separada, dê um nome claro ao que a função faz, e tente usar sempre somente 1 linha dentro do if.

Isso vai ajudar a legibilidade do código =)

**Comment by alanhr** [comment]
Eu acho que quando você começa a ter muitos IF/ELSE você tem que começar  aplicar um pattern, porque dar manutenção em um código com muitos  IF/ELSE é imposivel

**Comment by fdaciuk** [comment]
Pois é @alanhr.. se for pra usar um if ou outro, tudo bem. Mas se for pra ter um if gigante, existem algumas soluções. Escrevi um post sobre isso ano passado: http://blog.da2k.com.br/2015/01/22/clean-code-em-javascript-troque-seu-if-gigante-por-metodos/

**Comment by alanhr** [comment]
@fdaciuk cara eu vejo muito código cheio de if e else, eu as vezes acho que a galera esquece de estudar patterns

**Comment by suissa** [comment]
#dica Object Calesthenics

Se vc usa funções puras ou atômicas facilmente vc elimina o else, pois no if vc dá o return logo o q vier depois é o else.
